solidity/contracts/liquidity-protection/TokenTimeWeightedAverage.sol

```diff
@@ -0,0 +1,312 @@
+// SPDX-License-Identifier: SEE LICENSE IN LICENSE
+pragma solidity 0.6.12;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+
+import "./interfaces/ITokenTimeWeightedAverage.sol";
+
+import "../utility/Types.sol";
+import "../utility/SafeMath.sol";
+import "../utility/Math.sol";
+import "../utility/Utils.sol";
+
+/**
+ * @dev Token Time-Weighted Average contract
+ */
+contract TokenTimeWeightedAverage is ITokenTimeWeightedAverage, AccessControl, Utils {
+    using SafeMath for uint256;
+
+    struct TokenData {
+        uint256 firstSampleTime;
+        uint256 lastSampleTime;
+        uint256 prevAccumulatorTime;
+        Fraction prevAccumulator;
+        mapping(uint256 => Fraction) accumulators;
+        mapping(uint256 => bool) timestamps;
+    }
+
+    uint256 private constant INITIAL_SAMPLE_N = 0;
+    uint256 private constant INITIAL_SAMPLE_D = 1;
+    uint256 private constant MAX_UINT128 = 2**128 - 1;
```

Comment:
Missing spaces before and after `**` (probably my fault in the original commit).

Reply:
Prettier... nothing to do about that, other than to chillax.

Reply:
+1 for chillax